core/src/main/java/org/elasticsearch/index/shard/TranslogRecoveryPerformer.java

```diff
@@ -146,14 +146,21 @@ private void maybeAddMappingUpdate(String type, Mapping update, String docId, bo
      *                            cause a {@link MapperException} to be thrown if an update
      *                            is encountered.
      */
-    public void performRecoveryOperation(Engine engine, Translog.Operation operation, boolean allowMappingUpdates) {
+    private void performRecoveryOperation(Engine engine, Translog.Operation operation, boolean allowMappingUpdates, Engine.Operation.Origin origin) {
+        performRecoveryOperation(engine, operation, null, allowMappingUpdates, origin);
+    }
+
+    private void performRecoveryOperation(final Engine engine, final Translog.Operation operation, final Translog.Location location, final boolean allowMappingUpdates, final Engine.Operation.Origin origin) {
         try {
             switch (operation.opType()) {
                 case INDEX:
                     Translog.Index index = (Translog.Index) operation;
                     Engine.Index engineIndex = IndexShard.prepareIndex(docMapper(index.type()), source(shardId.getIndexName(), index.type(), index.id(), index.source())
-                                    .routing(index.routing()).parent(index.parent()).timestamp(index.timestamp()).ttl(index.ttl()),
-                            index.version(), index.versionType().versionTypeForReplicationAndRecovery(), Engine.Operation.Origin.RECOVERY);
+                            .routing(index.routing()).parent(index.parent()).timestamp(index.timestamp()).ttl(index.ttl()),
+                        index.version(), index.versionType().versionTypeForReplicationAndRecovery(), origin);
+                    if (location != null) {
+                        engineIndex.setTranslogLocation(location);
```

Comment:
The indexing memory controller listens to operations and starts the accounting sequence based on the size from the translog operation.